src/stream/src/executor/wrapper/trace.rs

```diff
@@ -17,50 +17,58 @@ use std::sync::Arc;
 use await_tree::InstrumentAwait;
 use futures::{pin_mut, StreamExt};
 use futures_async_stream::try_stream;
-use tracing::{event, Instrument};
+use tracing::{Instrument, Span};
 
 use crate::executor::error::StreamExecutorError;
 use crate::executor::monitor::StreamingMetrics;
 use crate::executor::{ExecutorInfo, Message, MessageStream};
 use crate::task::ActorId;
 
-/// Streams wrapped by `trace` will print data passing in the stream graph to stdout.
+/// Streams wrapped by `trace` will be traced with `tracing` spans and reported to `opentelemetry`.
 #[try_stream(ok = Message, error = StreamExecutorError)]
 pub async fn trace(
     info: Arc<ExecutorInfo>,
-    input_pos: usize,
+    _input_pos: usize,
     actor_id: ActorId,
-    _executor_id: u64,
+    executor_id: u64,
     metrics: Arc<StreamingMetrics>,
     input: impl MessageStream,
 ) {
-    let span_name = format!("{}_{}_next", info.identity, input_pos);
     let actor_id_string = actor_id.to_string();
 
+    let span_name = pretty_identity(&info.identity, actor_id, executor_id);
+
+    let new_span = || tracing::info_span!("executor", "otel.name" = span_name, actor_id);
+    let mut span = new_span();
+
     pin_mut!(input);
 
-    while let Some(message) = input
-        .next()
-        .instrument(tracing::trace_span!(
-            "next",
-            otel.name = span_name.as_str(),
-            next = info.identity,
-            input_pos = input_pos
-        ))
-        .await
-        .transpose()?
-    {
+    while let Some(message) = input.next().instrument(span.clone()).await.transpose()? {
         if let Message::Chunk(chunk) = &message {
             if chunk.cardinality() > 0 {
                 metrics
                     .executor_row_count
                     .with_label_values(&[&actor_id_string, &info.identity])
                     .inc_by(chunk.cardinality() as u64);
-                event!(tracing::Level::TRACE, prev = %info.identity, msg = "chunk", "input = \n{:#?}", chunk);
+                tracing::trace!(prev = %info.identity, msg = "chunk", "input = \n{:#?}", chunk);
             }
         }
 
-        yield message;
+        match &message {
+            Message::Chunk(_) | Message::Watermark(_) => yield message,
+
+            Message::Barrier(_barrier) => {
+                // Drop the span as the inner executor has finished processing the barrier (then all
+                // data from the previous epoch).
+                let _ = std::mem::replace(&mut span, Span::none());
+
+                yield message;
+
+                // Create a new span after we're called again. Now we're in a new epoch and the
+                // parent of the span is updated.
+                span = new_span();
```

Comment:
The ordering here could be really subtle. I find the current implementation to be the most understandable.